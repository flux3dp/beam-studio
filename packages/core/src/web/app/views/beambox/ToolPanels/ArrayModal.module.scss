.modal {
  .close-icon {
    font-size: 22px;
  }
  .title {
    text-align: center;
    font-size: 17px;
    font-weight: 600;
    margin: 12px 0;
  }
  .subtitle {
    font-weight: 600;
  }
  .field {
    display: flex;
    justify-content: space-between;
    align-items: center;
    column-gap: 8px;
    .label {
      min-width: 60px;
    }
    .slider {
      flex-grow: 1;
    }
    .input {
      width: 90px;
      input {
        border-radius: 0;
      }
      &.with-unit {
        flex-direction: row-reverse;
        column-gap: 4px;
        .unit {
          color: #7c7c7c;
        }
      }
    }
  }
}
