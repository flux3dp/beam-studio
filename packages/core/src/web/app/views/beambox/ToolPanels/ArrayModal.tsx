import classNames from 'classnames';
import React from 'react';
import Icon from '@ant-design/icons';
import { ConfigProvider, InputNumber, Modal, Slider } from 'antd';

import ActionPanelIcons from 'app/icons/action-panel/ActionPanelIcons';
import i18n from 'helpers/i18n';
import storage from 'implementations/storage';
import units from 'helpers/units';

import styles from './ArrayModal.module.scss';

const LANG = i18n.lang.beambox.tool_panels;
const unitSettings: {
  [key: string]: {
    precision: number;
    distance: { default: number; min: number; max: number; step?: number };
  };
} = {
  mm: {
    precision: 2,
    distance: { default: 20, min: 1, max: 50 },
  },
  inch: {
    precision: 2,
    distance: { default: 1, min: 0, max: 2, step: 0.1 },
  },
};

interface Value {
  row: number;
  column: number;
  dx: number;
  dy: number;
}
interface Props {
  onCancel: () => void;
  onOk: (data: Value) => void;
}

const ArrayModal = ({ onCancel, onOk }: Props): JSX.Element => {
  const unit = React.useMemo(() => (storage.get('default-units') === 'inches' ? 'inch' : 'mm'), []);
  const setting = unitSettings[unit];
  const [data, setData] = React.useState<Value>({
    row: 3,
    column: 3,
    dx: setting.distance.default,
    dy: setting.distance.default,
  });
  const { row, column, dx, dy } = data;
  const setRow = (val) => setData({ ...data, row: val });
  const setColumn = (val) => setData({ ...data, column: val });
  const setDx = (val) => setData({ ...data, dx: val });
  const setDy = (val) => setData({ ...data, dy: val });
  return (
    <ConfigProvider
      theme={{
        components: {
          Button: { borderRadius: 100 },
          InputNumber: { borderRadiusSM: 100 },
        },
      }}
    >
      <Modal
        className={styles.modal}
        closeIcon={<Icon className={styles['close-icon']} component={ActionPanelIcons.Delete} />}
        okText={LANG.confirm}
        cancelText={LANG.cancel}
        onOk={() => {
          const dxInMM = +units.convertUnit(data.dx, 'mm', unit).toFixed(2);
          const dyInMM = +units.convertUnit(data.dy, 'mm', unit).toFixed(2);
          onOk({ ...data, dx: dxInMM, dy: dyInMM });
        }}
        onCancel={onCancel}
        centered
        open
      >
        <div className={styles.title}>{LANG.grid_array}</div>
        {/* TODO: add preview */}
        <div className={styles.subtitle}>{LANG.array_dimension}</div>
        <div className={styles.field}>
          <span className={styles.label}>{LANG.columns}</span>
          <Slider className={styles.slider} min={1} max={10} value={column} onChange={setColumn} />
          <InputNumber
            className={styles.input}
            type="number"
            size="small"
            min={1}
            precision={0}
            value={column}
            onChange={setColumn}
            controls={false}
          />
        </div>
        <div className={styles.field}>
          <span className={styles.label}>{LANG.rows}</span>
          <Slider className={styles.slider} min={1} max={10} value={row} onChange={setRow} />
          <InputNumber
            className={styles.input}
            type="number"
            size="small"
            min={1}
            precision={0}
            value={row}
            onChange={setRow}
            controls={false}
          />
        </div>
        <div className={styles.subtitle}>{LANG.array_interval}</div>
        <div className={styles.field}>
          <span className={styles.label}>{LANG.dx}</span>
          <Slider
            className={styles.slider}
            min={setting.distance.min}
            max={setting.distance.max}
            step={setting.distance.step}
            value={dx}
            onChange={setDx}
          />
          <InputNumber
            className={classNames(styles.input, styles['with-unit'])}
            type="number"
            size="small"
            min={0}
            precision={setting.precision}
            value={dx}
            onChange={setDx}
            prefix={<span className={styles.unit}>{unit}</span>}
            controls={false}
          />
        </div>
        <div className={styles.field}>
          <span className={styles.label}>{LANG.dy}</span>
          <Slider
            className={styles.slider}
            min={setting.distance.min}
            max={setting.distance.max}
            step={setting.distance.step}
            value={dy}
            onChange={setDy}
          />
          <InputNumber
            className={classNames(styles.input, styles['with-unit'])}
            type="number"
            size="small"
            min={0}
            precision={setting.precision}
            value={dy}
            onChange={setDy}
            prefix={<span className={styles.unit}>{unit}</span>}
            controls={false}
          />
        </div>
      </Modal>
    </ConfigProvider>
  );
};
export default ArrayModal;
