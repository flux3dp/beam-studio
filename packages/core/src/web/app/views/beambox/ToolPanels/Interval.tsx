import classNames from 'classnames';
import React from 'react';

import BeamboxPreference from 'app/actions/beambox/beambox-preference';
import i18n from 'helpers/i18n';
import storage from 'implementations/storage';
import UnitInput from 'app/widgets/Unit-Input-v2';
import { getWorkarea } from 'app/constants/workarea-constants';

const LANG = i18n.lang.beambox.tool_panels;

interface Props {
  dx?: number,
  dy?: number,
  onValueChange?: (rc: { dx: number, dy: number }) => void,
}

interface State {
  dx: number,
  dy: number,
  isCollapsed: boolean,
}

class Interval extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    const { dx, dy } = this.props;
    this.state = {
      dx,
      dy,
      isCollapsed: false,
    };
  }

  UNSAFE_componentWillReceiveProps(nextProps: Props): void {
    this.setState({
      dx: nextProps.dx,
      dy: nextProps.dy,
    });
  }

  onDxChanged = (dx: number): void => {
    const { onValueChange } = this.props;
    const { dy } = this.state;
    onValueChange({
      dx,
      dy,
    });
    this.setState({ dx });
  };

  onDyChanged = (dy: number): void => {
    const { onValueChange } = this.props;
    const { dx } = this.state;
    onValueChange({
      dx,
      dy,
    });
    this.setState({ dy });
  };

  getValueCaption = (): string => {
    const { dx, dy } = this.state;
    const units = storage.get('default-units') || 'mm';
    return units === 'inches'
      ? `${Number(dx / 25.4).toFixed(3)}", ${Number(dy / 25.4).toFixed(3)}"`
      : `${dx}, ${dy} mm`;
  };

  render(): JSX.Element {
    const { dx, dy, isCollapsed } = this.state;
    const workarea = getWorkarea(BeamboxPreference.read('workarea'));
    return (
      <div className="tool-panel">
        <label className="controls accordion">
          <input type="checkbox" className="accordion-switcher" defaultChecked />
          <p className="caption" onClick={() => this.setState({ isCollapsed: !isCollapsed })}>
            {LANG.array_interval}
            <span className="value">{this.getValueCaption()}</span>
          </p>
          <div className={classNames('tool-panel-body', { collapsed: isCollapsed })}>
            <div className="control">
              <span className="text-center header">{LANG.dx}</span>
              <UnitInput
                id="array_width"
                min={0}
                max={workarea.width}
                unit="mm"
                defaultValue={dx}
                getValue={this.onDxChanged}
              />
            </div>
            <div className="control">
              <span className="text-center header">{LANG.dy}</span>
              <UnitInput
                id="array_height"
                min={0}
                max={workarea.displayHeight || workarea.height}
                unit="mm"
                defaultValue={dy}
                getValue={this.onDyChanged}
              />
            </div>
          </div>
        </label>
      </div>
    );
  }
}

export default Interval;
