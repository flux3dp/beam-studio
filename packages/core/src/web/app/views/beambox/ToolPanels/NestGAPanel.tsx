import React from 'react';

import UnitInput from 'app/widgets/Unit-Input-v2';

interface Props {
  nestOptions: {
    generations: number,
    population: number,
  }
  updateNestOptions: (options: {
    generations?: number,
    population?: number,
  }) => void,
}

function NestGAPanel({ nestOptions, updateNestOptions }: Props): JSX.Element {
  const [generations, updateGenerations] = React.useState(nestOptions.generations);
  const [population, updatePopulation] = React.useState(nestOptions.population);

  const updateGen = (val: number) => {
    updateNestOptions({ generations: val });
    updateGenerations(val);
  };

  const updatePopu = (val: number) => {
    updateNestOptions({ population: val });
    updatePopulation(val);
  };

  return (
    <div className="tool-panel">
      <label className="controls accordion">
        <input type="checkbox" className="accordion-switcher" defaultChecked />
        <p className="caption">
          GA
          <span className="value">{`G${generations}, P${population}`}</span>
        </p>
        <label className="accordion-body">
          <div>
            <span className="text-center header">Generations</span>
            <div className="control">
              <UnitInput
                min={1}
                unit=""
                decimal={0}
                defaultValue={generations}
                getValue={updateGen}
              />
            </div>
            <span className="text-center header">Population</span>
            <div className="control">
              <UnitInput
                min={2}
                unit=""
                decimal={0}
                defaultValue={population}
                getValue={updatePopu}
              />
            </div>
          </div>
        </label>
      </label>
    </div>
  );
}

export default NestGAPanel;
