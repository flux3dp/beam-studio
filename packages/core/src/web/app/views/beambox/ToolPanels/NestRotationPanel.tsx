import React from 'react';

import UnitInput from '@core/app/widgets/Unit-Input-v2';
import i18n from '@core/helpers/i18n';

const LANG = i18n.lang.beambox.tool_panels;

interface Props {
  onValueChange: (val: number) => void;
  rotations: number;
}

function NestRotationPanel({ onValueChange, rotations: rotationsProps }: Props): React.JSX.Element {
  const [rotations, updateRotations] = React.useState(rotationsProps);

  const updateVal = (val) => {
    onValueChange(val);
    updateRotations(val);
  };

  return (
    <div className="tool-panel">
      <label className="controls accordion">
        <input className="accordion-switcher" defaultChecked type="checkbox" />
        <p className="caption">
          {LANG._nest.rotations}
          <span className="value">{rotations.toString()}</span>
        </p>
        <label className="accordion-body">
          <div>
            <div className="control nest-rotations">
              <UnitInput decimal={0} defaultValue={rotations} getValue={updateVal} min={1} unit="" />
            </div>
          </div>
        </label>
      </label>
    </div>
  );
}

export default NestRotationPanel;
