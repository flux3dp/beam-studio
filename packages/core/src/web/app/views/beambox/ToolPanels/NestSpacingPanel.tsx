import React from 'react';

import i18n from 'helpers/i18n';
import storage from 'implementations/storage';
import UnitInput from 'app/widgets/Unit-Input-v2';

const LANG = i18n.lang.beambox.tool_panels;

interface Props {
  spacing: number,
  onValueChange: (val: number) => void,
}

function NestSpacingPanel({ spacing: spacingProps, onValueChange }: Props): JSX.Element {
  const [spacing, updateSpacing] = React.useState(spacingProps);

  const updateVal = (val) => {
    onValueChange(val);
    updateSpacing(val);
  };

  const getValueCaption = () => {
    const units = storage.get('default-units') || 'mm';
    return units === 'inches' ? `${Number(spacing / 25.4).toFixed(3)}\"` : `${spacing} mm`;
  };

  return (
    <div className="tool-panel">
      <label className="controls accordion">
        <input type="checkbox" className="accordion-switcher" defaultChecked />
        <p className="caption">
          {LANG._nest.spacing}
          <span className="value">{getValueCaption()}</span>
        </p>
        <label className="accordion-body">
          <div>
            <div className="control nest-spacing">
              <UnitInput
                min={0}
                unit="mm"
                defaultValue={spacing}
                getValue={(val) => updateVal(val)}
              />
            </div>
          </div>
        </label>
      </label>
    </div>
  );
}

export default NestSpacingPanel;
