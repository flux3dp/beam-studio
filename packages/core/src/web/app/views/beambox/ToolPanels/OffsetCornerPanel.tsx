import classNames from 'classnames';
import React from 'react';

import i18n from 'helpers/i18n';
import SelectView from 'app/widgets/Select';

const LANG = i18n.lang.beambox.tool_panels;

interface Props {
  cornerType: string,
  onValueChange: (val: string) => void,
}

function OffsetCornerPanel({ cornerType: cornerTypeProps, onValueChange }: Props): JSX.Element {
  const [cornerType, updateCornerType] = React.useState(cornerTypeProps);
  const [isCollapsed, updateIsCollapsed] = React.useState(false);

  const updateOffsetCorner = (val: string) => {
    onValueChange(val);
    updateCornerType(val);
  };

  const getOffsetCornerText = () => ({
    sharp: LANG._offset.sharp,
    round: LANG._offset.round,
  }[cornerType]);

  const options = [
    {
      value: 'sharp',
      label: LANG._offset.sharp,
      selected: cornerType === 'sharp',
    },
    {
      value: 'round',
      label: LANG._offset.round,
      selected: cornerType === 'round',
    },
  ];
  return (
    <div className="tool-panel">
      <label className="controls accordion">
        <input type="checkbox" className="accordion-switcher" defaultChecked />
        <p className="caption" onClick={() => updateIsCollapsed(!isCollapsed)}>
          {LANG._offset.corner_type}
          <span className="value">{getOffsetCornerText()}</span>
        </p>
        <div className={classNames('tool-panel-body', { collapsed: isCollapsed })}>
          <div className="control offset-corner">
            <SelectView
              id="select-offset-corner"
              options={options}
              onChange={(e) => updateOffsetCorner(e.target.value)}
            />
          </div>
        </div>
      </label>
    </div>
  );
}

export default OffsetCornerPanel;
