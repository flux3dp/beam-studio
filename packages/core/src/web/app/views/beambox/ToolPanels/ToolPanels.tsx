import React from 'react';

import classNames from 'classnames';

import Constant from '@core/app/actions/beambox/constant';
import type { ToolPanelType } from '@core/app/actions/beambox/toolPanelsController';
import Dialog from '@core/app/actions/dialog-caller';
import currentFileManager from '@core/app/svgedit/currentFileManager';
import ObjectPanelController from '@core/app/views/beambox/Right-Panels/contexts/ObjectPanelController';
import ArrayModal from '@core/app/views/beambox/ToolPanels/ArrayModal';
import IntervalPanel from '@core/app/views/beambox/ToolPanels/Interval';
import NestGAPanel from '@core/app/views/beambox/ToolPanels/NestGAPanel';
import NestRotationPanel from '@core/app/views/beambox/ToolPanels/NestRotationPanel';
import NestSpacingPanel from '@core/app/views/beambox/ToolPanels/NestSpacingPanel';
import OffsetCornerPanel from '@core/app/views/beambox/ToolPanels/OffsetCornerPanel';
import OffsetDirectionPanel from '@core/app/views/beambox/ToolPanels/OffsetDirectionPanel';
import OffsetDistancePanel from '@core/app/views/beambox/ToolPanels/OffsetDistancePanel';
import OffsetModal from '@core/app/views/beambox/ToolPanels/OffsetModal';
import RowColumnPanel from '@core/app/views/beambox/ToolPanels/RowColumn';
import offsetElements from '@core/helpers/clipper/offset';
import eventEmitterFactory from '@core/helpers/eventEmitterFactory';
import i18n from '@core/helpers/i18n';
import { getSVGAsync } from '@core/helpers/svg-editor-helper';
import { isMobile } from '@core/helpers/system-helper';
import storage from '@core/implementations/storage';

let svgCanvas;

getSVGAsync((globalSVG) => {
  svgCanvas = globalSVG.Canvas;
});

const drawingToolEventEmitter = eventEmitterFactory.createEventEmitter('drawing-tool');

const LANG = i18n.lang.beambox.tool_panels;

const _mm2pixel = (pixel_input) => {
  const { dpmm } = Constant;

  return Number(pixel_input * dpmm);
};

const validPanelsMap = {
  gridArray: ['rowColumn', 'distance'],
  nest: ['nestOffset', 'nestRotation', 'nestGA'],
  offset: ['offsetDir', 'offsetCorner', 'offsetDist'],
  unknown: [],
};

interface Props {
  data: any;
  type: ToolPanelType;
  unmount: () => void;
}

class ToolPanel extends React.Component<Props> {
  private offset: {
    cornerType: string;
    dir: number;
    distance: number;
  };

  private nestOptions: {
    generations: number;
    population: number;
    rotations: number;
    spacing: number;
  };

  constructor(props) {
    super(props);
    this._setArrayRowColumn = this._setArrayRowColumn.bind(this);
    this._setArrayDistance = this._setArrayDistance.bind(this);
    this._setOffsetDir = this._setOffsetDir.bind(this);
    this._setOffsetCorner = this._setOffsetCorner.bind(this);
    this._setOffsetDist = this._setOffsetDist.bind(this);
    this.offset = {
      cornerType: 'sharp',
      dir: 1, // 1 for outward, 0 for inward
      distance: 5,
    };
    this.nestOptions = {
      generations: 3,
      population: 10,
      rotations: 1,
      spacing: 0,
    };
  }

  _setArrayRowColumn(rowcolumn) {
    this.props.data.rowcolumn = rowcolumn;
    this.setState({ rowcolumn });
  }

  _setArrayDistance(distance) {
    this.props.data.distance = distance;
    this.setState({ distance });
  }

  _setOffsetDir(dir) {
    this.offset.dir = dir;
  }

  _setOffsetDist(val) {
    this.offset.distance = val;
  }

  _setOffsetCorner(val) {
    this.offset.cornerType = val;
  }

  renderPanels() {
    const { data, type } = this.props;
    const validPanels = validPanelsMap[type] || validPanelsMap.unknown;
    const panelsToBeRender = [];

    for (let i = 0; i < validPanels.length; ++i) {
      const panelName = validPanels[i];
      let panel;

      switch (panelName) {
        case 'rowColumn':
          panel = <RowColumnPanel key={panelName} {...data.rowcolumn} onValueChange={this._setArrayRowColumn} />;
          break;
        case 'distance':
          panel = <IntervalPanel key={panelName} {...data.distance} onValueChange={this._setArrayDistance} />;
          break;
        case 'offsetDir':
          panel = <OffsetDirectionPanel dir={this.offset.dir} key={panelName} onValueChange={this._setOffsetDir} />;
          break;
        case 'offsetCorner':
          panel = (
            <OffsetCornerPanel
              cornerType={this.offset.cornerType}
              key={panelName}
              onValueChange={this._setOffsetCorner}
            />
          );
          break;
        case 'offsetDist':
          panel = (
            <OffsetDistancePanel distance={this.offset.distance} key={panelName} onValueChange={this._setOffsetDist} />
          );
          break;
        case 'nestOffset':
          panel = (
            <NestSpacingPanel
              key={panelName}
              onValueChange={(val) => {
                this.nestOptions.spacing = val;
              }}
              spacing={this.nestOptions.spacing}
            />
          );
          break;
        case 'nestGA':
          panel = (
            <NestGAPanel
              key={panelName}
              nestOptions={this.nestOptions}
              updateNestOptions={(options) => {
                this.nestOptions = { ...this.nestOptions, ...options };
              }}
            />
          );
          break;
        case 'nestRotation':
          panel = (
            <NestRotationPanel
              key={panelName}
              onValueChange={(val) => {
                this.nestOptions.rotations = val;
              }}
              rotations={this.nestOptions.rotations}
            />
          );
          break;
        default:
          break;
      }
      panelsToBeRender.push(panel);
    }

    return panelsToBeRender;
  }

  renderTitle() {
    const { type } = this.props;
    const titleMap = {
      gridArray: LANG.grid_array,
      offset: LANG.offset,
    };
    const title = titleMap[type];

    return (
      <div className="tool-panel">
        <label className="controls accordion">
          <input className="accordion-switcher" defaultChecked type="checkbox" />
          <p className="caption">
            <span className="value">{title}</span>
          </p>
        </label>
      </div>
    );
  }

  _onCancel = () => {
    this.props.unmount();
    svgCanvas.setMode('select');
    drawingToolEventEmitter.emit('SET_ACTIVE_BUTTON', 'Cursor');
  };

  _getOnYes = (): (() => Promise<void> | void) => {
    const { data, type, unmount } = this.props;

    switch (type) {
      case 'gridArray':
        return async () => {
          const distance = {
            dx: _mm2pixel(data.distance.dx),
            dy: _mm2pixel(data.distance.dy),
          };

          await svgCanvas.gridArraySelectedElement(distance, data.rowcolumn);
          unmount();
          svgCanvas.setMode('select');
          drawingToolEventEmitter.emit('SET_ACTIVE_BUTTON', 'Cursor');
          currentFileManager.setHasUnsavedChanges(true);
        };

      case 'offset':
        return () => {
          offsetElements(this.offset.dir, _mm2pixel(this.offset.distance), this.offset.cornerType as 'round' | 'sharp');
          unmount();
          svgCanvas.setMode('select');
          drawingToolEventEmitter.emit('SET_ACTIVE_BUTTON', 'Cursor');
          currentFileManager.setHasUnsavedChanges(true);
        };
      case 'nest':
        return () => {
          this.nestOptions.spacing *= 10; // pixel to mm
          svgCanvas.nestElements(null, null, this.nestOptions);
          unmount();
          svgCanvas.setMode('select');
          drawingToolEventEmitter.emit('SET_ACTIVE_BUTTON', 'Cursor');
        };
      default:
        return () => unmount();
    }
  };

  renderButtons() {
    return (
      <div className="tool-block">
        <div className="btn-h-group">
          <button className="btn btn-default primary" onClick={this._getOnYes()} type="button">
            {LANG.confirm}
          </button>
          <button className="btn btn-default" onClick={this._onCancel} type="button">
            {LANG.cancel}
          </button>
        </div>
      </div>
    );
  }

  renderModal(): void {
    const { type, unmount } = this.props;

    if (Dialog.isIdExist(type)) {
      return;
    }

    const closeModal = () => {
      ObjectPanelController.updateActiveKey(null);
      Dialog.popDialogById(type);
    };
    const onOk = this._getOnYes();
    const onCancel = this._onCancel;

    switch (type) {
      case 'gridArray':
        Dialog.addDialogComponent(
          type,
          <ArrayModal
            onCancel={() => {
              onCancel();
              closeModal();
            }}
            onOk={async (value) => {
              this._setArrayRowColumn({ column: value.column, row: value.row });
              this._setArrayDistance({ dx: value.dx, dy: value.dy });
              await onOk();
              closeModal();
            }}
          />,
        );
        break;
      case 'offset':
        Dialog.addDialogComponent(
          type,
          <OffsetModal
            onCancel={() => {
              onCancel();
              closeModal();
            }}
            onOk={async (value) => {
              this._setOffsetDir(value.dir);
              this._setOffsetDist(value.distance);
              this._setOffsetCorner(value.cornerType);
              await onOk();
              closeModal();
            }}
          />,
        );
        break;
      default:
        ObjectPanelController.updateActiveKey(null);
        unmount();
    }
  }

  _findPositionStyle() {
    return {
      bottom: 10,
      left: 50,
      position: 'absolute',
      zIndex: 999,
    } as React.CSSProperties;
  }

  render() {
    if (isMobile()) {
      this.renderModal();

      return null;
    }

    const lang = storage.get('active-lang') || 'en';
    const positionStyle = this._findPositionStyle();
    const classes = classNames('tool-panels', lang);

    return (
      <div className={classes} id="beamboxToolPanel" style={positionStyle}>
        {this.renderTitle()}
        {this.renderPanels()}
        {this.renderButtons()}
      </div>
    );
  }
}

export default ToolPanel;
