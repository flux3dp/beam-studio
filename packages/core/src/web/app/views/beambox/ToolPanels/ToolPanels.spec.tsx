import React from 'react';

import { render } from '@testing-library/react';

const get = jest.fn();

jest.mock('@core/implementations/storage', () => ({
  get,
}));

const getSVGAsync = jest.fn();

jest.mock('@core/helpers/svg-editor-helper', () => ({
  getSVGAsync,
}));

const setMode = jest.fn();
const gridArraySelectedElement = jest.fn();
const nestElements = jest.fn();

getSVGAsync.mockImplementation((callback) => {
  callback({
    Canvas: {
      gridArraySelectedElement,
      nestElements,
      setMode,
    },
  });
});

const mockSetHasUnsavedChanges = jest.fn();

jest.mock('@core/app/svgedit/currentFileManager', () => ({
  setHasUnsavedChanges: (...args) => mockSetHasUnsavedChanges(...args),
}));

const offsetElements = jest.fn();

jest.mock(
  '@core/helpers/clipper/offset',
  () =>
    (...args) =>
      offsetElements(...args),
);

jest.mock('@core/app/views/beambox/ToolPanels/Interval', () => 'dummy-interval');
jest.mock('@core/app/views/beambox/ToolPanels/NestGAPanel', () => ({ nestOptions }: any) => (
  <div>dummy-nest-ga-panel nestOptions: {JSON.stringify(nestOptions)}</div>
));
jest.mock('@core/app/views/beambox/ToolPanels/NestRotationPanel', () => 'dummy-nest-rotation-panel');
jest.mock('@core/app/views/beambox/ToolPanels/NestSpacingPanel', () => 'dummy-nest-spacing-panel');
jest.mock('@core/app/views/beambox/ToolPanels/OffsetPanel', () => 'dummy-offset-panel');
jest.mock('@core/app/views/beambox/ToolPanels/RowColumn', () => 'dummy-row-column');
jest.mock('@core/app/actions/beambox/toolPanelsController', () => ({}));

const isMobile = jest.fn();

jest.mock('@core/helpers/system-helper', () => ({
  isMobile: () => isMobile(),
}));

const isIdExist = jest.fn();
const popDialogById = jest.fn();
const addDialogComponent = jest.fn();

jest.mock('@core/app/actions/dialog-caller', () => ({
  addDialogComponent,
  isIdExist,
  popDialogById,
}));

import ToolPanels from './ToolPanels';

describe('should render correctly', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('type is gridArray', () => {
    const unmount = jest.fn();
    const { container } = render(
      <ToolPanels
        data={{
          distance: {
            dx: 0,
            dy: 0,
          },
          rowcolumn: {
            column: 1,
            row: 1,
          },
        }}
        type="gridArray"
        unmount={unmount}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('type is offset', () => {
    const unmount = jest.fn();
    const { container } = render(
      <ToolPanels
        data={{
          distance: {
            dx: 0,
            dy: 0,
          },
          rowcolumn: {
            column: 1,
            row: 1,
          },
        }}
        type="offset"
        unmount={unmount}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('type is nest', () => {
    const unmount = jest.fn();
    const { container } = render(
      <ToolPanels
        data={{
          distance: {
            dx: 0,
            dy: 0,
          },
          rowcolumn: {
            column: 1,
            row: 1,
          },
        }}
        type="nest"
        unmount={unmount}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});

describe('should render correctly in mobile', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    isMobile.mockReturnValue(true);
  });

  test('type is gridArray', () => {
    const unmount = jest.fn();
    const { container } = render(
      <ToolPanels
        data={{
          distance: {
            dx: 0,
            dy: 0,
          },
          rowcolumn: {
            column: 1,
            row: 1,
          },
        }}
        type="gridArray"
        unmount={unmount}
      />,
    );

    expect(container).toMatchSnapshot();
    expect(isIdExist).toBeCalledTimes(1);
    expect(isIdExist).toBeCalledWith('gridArray');
    expect(addDialogComponent).toBeCalledTimes(1);
    expect(addDialogComponent).toBeCalledWith('gridArray', expect.anything());
    expect(addDialogComponent.mock.calls[0][1]).toMatchSnapshot();
  });

  test('type is offset', () => {
    const unmount = jest.fn();
    const { container } = render(
      <ToolPanels
        data={{
          distance: {
            dx: 0,
            dy: 0,
          },
          rowcolumn: {
            column: 1,
            row: 1,
          },
        }}
        type="offset"
        unmount={unmount}
      />,
    );

    expect(container).toMatchSnapshot();
    expect(isIdExist).toBeCalledWith('offset');
    expect(addDialogComponent).toBeCalledTimes(1);
    expect(addDialogComponent).toBeCalledWith('offset', expect.anything());
    expect(addDialogComponent.mock.calls[0][1]).toMatchSnapshot();
  });
});
