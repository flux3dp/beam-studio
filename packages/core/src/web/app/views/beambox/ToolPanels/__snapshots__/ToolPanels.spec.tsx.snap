// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly in mobile type is gridArray 1`] = `<div />`;

exports[`should render correctly in mobile type is gridArray 2`] = `
<ArrayModal
  onCancel={[Function]}
  onOk={[Function]}
/>
`;

exports[`should render correctly in mobile type is offset 1`] = `<div />`;

exports[`should render correctly in mobile type is offset 2`] = `
<OffsetModal
  onCancel={[Function]}
  onOk={[Function]}
/>
`;

exports[`should render correctly type is gridArray 1`] = `
<div>
  <div
    class="tool-panels en"
    id="beamboxToolPanel"
    style="bottom: 10px; left: 50px; position: absolute; z-index: 999;"
  >
    <div
      class="tool-panel"
    >
      <label
        class="controls accordion"
      >
        <input
          checked=""
          class="accordion-switcher"
          type="checkbox"
        />
        <p
          class="caption"
        >
          <span
            class="value"
          >
            Create Grid Array
          </span>
        </p>
      </label>
    </div>
    <dummy-row-column
      column="1"
      row="1"
    />
    <dummy-interval
      dx="0"
      dy="0"
    />
    <div
      class="tool-block"
    >
      <div
        class="btn-h-group"
      >
        <button
          class="btn btn-default primary"
          type="button"
        >
          Confirm
        </button>
        <button
          class="btn btn-default"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly type is nest 1`] = `
<div>
  <div
    class="tool-panels en"
    id="beamboxToolPanel"
    style="bottom: 10px; left: 50px; position: absolute; z-index: 999;"
  >
    <div
      class="tool-panel"
    >
      <label
        class="controls accordion"
      >
        <input
          checked=""
          class="accordion-switcher"
          type="checkbox"
        />
        <p
          class="caption"
        >
          <span
            class="value"
          />
        </p>
      </label>
    </div>
    <dummy-nest-spacing-panel
      spacing="0"
    />
    <dummy-nest-rotation-panel
      rotations="1"
    />
    <div>
      dummy-nest-ga-panel nestOptions: 
      {"generations":3,"population":10,"rotations":1,"spacing":0}
    </div>
    <div
      class="tool-block"
    >
      <div
        class="btn-h-group"
      >
        <button
          class="btn btn-default primary"
          type="button"
        >
          Confirm
        </button>
        <button
          class="btn btn-default"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly type is offset 1`] = `
<div>
  <div
    class="tool-panels en"
    id="beamboxToolPanel"
    style="bottom: 10px; left: 50px; position: absolute; z-index: 999;"
  >
    <div
      class="tool-panel"
    >
      <label
        class="controls accordion"
      >
        <input
          checked=""
          class="accordion-switcher"
          type="checkbox"
        />
        <p
          class="caption"
        >
          <span
            class="value"
          >
            Offset
          </span>
        </p>
      </label>
    </div>
    <dummy-offset-panel
      offset="[object Object]"
    />
    <div
      class="tool-block"
    >
      <div
        class="btn-h-group"
      >
        <button
          class="btn btn-default primary"
          type="button"
        >
          Confirm
        </button>
        <button
          class="btn btn-default"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
